### PR TITLE
Cuda graphs + loop labels RNN-T and TDT greedy decoding

### PR DESCRIPTION
**Pull request by artbataev**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add "CUDA graphs" implementation to label looping algorithm for greedy decoding (RNN-T and TDT)

# Usage

```shell
python examples/asr/speech_to_text_eval.py \
  pretrained_name=stt_en_fastconformer_transducer_large \
  dataset_manifest=test_other.json \
  batch_size=16 \
  output_filename=test_other_decoded.jsonl \
  amp=true \
  amp_dtype=bfloat16 \
  rnnt_decoding.greedy.loop_labels=true \
  rnnt_decoding.greedy.use_cuda_graph_decoder=true
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
jenkins

File: nemo/collections/asr/parts/submodules/rnnt_loop_labels_computer.py [hunk 1]
```diff
@@ -14,19 +14,179 @@
...
+
+try:
+    from cuda import cuda, cudart, nvrtc
+
+    HAVE_CUDA_PYTHON = True
+except ImportError:
+    HAVE_CUDA_PYTHON = False
+
+
+def run_nvrtc(kernel_string, kernel_name):
```
Comment (galv) [inline nemo/collections/asr/parts/submodules/rnnt_loop_labels_computer.py:40]: I have no problem with moving this copy-pasted function into `core/utils/cuda_python_utils.py` FWIW.

File: nemo/collections/asr/parts/submodules/rnnt_loop_labels_computer.py [hunk 2]
```diff
@@ -245,8 +419,331 @@ def __call__(
...
+
+        if torch.is_autocast_enabled():
+            encoder_output = encoder_output.to(torch.get_autocast_gpu_dtype())
+
+        # init or reinit graph
+        if self.state is None or self.state.need_reinit(encoder_output):
+            self._graph_reinitialize(encoder_output, encoder_output_length)
+
+        # copy (projected) encoder output and lenghts
+        self.state.encoder_output_projected[:current_batch_size, :current_max_time, ...].copy_(encoder_output)
```
Comment (galv) [inline nemo/collections/asr/parts/submodules/rnnt_loop_labels_computer.py:451]: Just a note for the future. If we get torch.while_loop to work with conditional nodes, we would be able to speed up these three torch ops by running them in three parallel cuda streams (I think this is called vertical or horizontal fusion), assuming torch inductor or thunder can do that right now.

**Review by galv — APPROVED** [review]
It looks good to me overall. I haven't run the code myself locally, but if I encounter weird performance issues later when I do, I can always make another PR for that.

**Comment by artbataev** [comment]
jenkins

File: nemo/collections/asr/modules/rnnt.py [hunk 3]
```diff
@@ -384,6 +384,13 @@ def batch_replace_states_mask(
         # same as `dst_states[0][mask] = src_states[0][mask]`, but non-blocking
         torch.where(mask.unsqueeze(-1), src_states[0], dst_states[0], out=dst_states[0])
 
+    @classmethod
+    def batch_replace_states_all(
+        cls, src_states: list[torch.Tensor], dst_states: list[torch.Tensor], mask: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/modules/rnnt.py]: Fixed

File: nemo/collections/asr/parts/utils/rnnt_utils.py [hunk 4]
```diff
@@ -557,10 +572,13 @@ def batched_hyps_to_hypotheses(
     Args:
         batched_hyps: BatchedHyps object
         alignments: BatchedAlignments object, optional; must correspond to BatchedHyps if present
+        batch_size: Batch Size to retrieve hypotheses. When working with CUDA graphs the batch size for all tensors
+            is constant, thus we need here the real batch size to return only necessary hypotheses
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/parts/utils/rnnt_utils.py:576]: Fixed

**Comment by artbataev** [comment]
jenkins

**Review by galv — DISMISSED** [review]

**Review by titu1994 — DISMISSED** [review]
Minor comments, awesome work !

File: nemo/collections/asr/parts/submodules/cuda_graph_rnnt_greedy_decoding.py [hunk 5]
```diff
@@ -104,96 +72,7 @@ def create_inner_while_loop_kernel():
...
-    yield body_stream, body_graph
-
-    cuda.cuLaunchKernel(
-        while_loop_kernel, 1, 1, 1, 1, 1, 1, 0, body_stream.cuda_stream, while_loop_args.ctypes.data, 0
-    )
-
-    cudart.cudaStreamEndCapture(body_stream.cuda_stream)
-
-    torch.cuda.set_stream(previous_stream)
+    return run_nvrtc(kernel_string, b"while_loop_conditional", _CUDA_PROGRAM_NAME)
```
[the remark this answers is not in the document]
Comment (galv) [reply nemo/collections/asr/parts/submodules/cuda_graph_rnnt_greedy_decoding.py:75]: cuda-python expects a byte string, not a normal python string. It is because cuda expects a `char *` in the underlying C API. Just a silly implementation detail.

File: nemo/collections/asr/modules/rnnt.py [hunk 6]
```diff
@@ -384,6 +384,13 @@ def batch_replace_states_mask(
         # same as `dst_states[0][mask] = src_states[0][mask]`, but non-blocking
         torch.where(mask.unsqueeze(-1), src_states[0], dst_states[0], out=dst_states[0])
 
+    @classmethod
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/modules/rnnt.py:387]: `AbstractRNNTDecoder` already has `batch_replace_states_all` method. 
`RNNTDecoder` and `StatelessTransducerDecoder` have different types of states: `tuple` for RNN-T and `list` for Stateless (but there is confusion in the code since, in most cases, the `list` type works, but not all).
I prefer to keep it clean and provide implementation for both classes. Also, due to planned work on the Transformer-Decoder, there will be a separate state type for the Transformer.

File: nemo/collections/asr/modules/rnnt_abstract.py [hunk 7]
```diff
@@ -285,6 +285,13 @@ def batch_replace_states_mask(
         """Replace states in dst_states with states from src_states using the mask, in a way that does not synchronize with the CPU"""
         raise NotImplementedError()
 
+    @classmethod
+    def batch_replace_states_all(
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/modules/rnnt_abstract.py:289]: There is nothing related to the `self`, so I prefer `classmethod` for this case. Also, I would prefer consistency with `batch_replace_states_mask`, which is also a `classmethod`.
For reusability purposes, `classmethod` is also better since if we know that this state is related to some class, we can use `classmethod` without requiring us to pass the actual decoder instance.

File: nemo/collections/asr/parts/submodules/rnnt_loop_labels_computer.py [hunk 8]
```diff
@@ -14,19 +14,148 @@
...
+        self.advance_mask_any = torch.tensor(True, device=self.device, dtype=torch.bool)
+
+        self.batched_hyps = rnnt_utils.BatchedHyps(
+            batch_size=self.batch_size,
+            init_length=self.max_time * max_symbols,
+            device=self.device,
+            float_dtype=float_dtype,
+        )
+        if preserve_alignments or preserve_frame_confidence:
+            self.alignments = rnnt_utils.BatchedAlignments(
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/parts/submodules/rnnt_loop_labels_computer.py:143]: It was already None, but I agree that it's better to indicate this explicitly. Added this to both classes

File: nemo/collections/asr/parts/submodules/rnnt_loop_labels_computer.py [hunk 9]
```diff
@@ -14,19 +14,148 @@
...
+    from cuda import cudart
+
+    HAVE_CUDA_PYTHON = True
+except ImportError:
+    HAVE_CUDA_PYTHON = False
+
+
+class LoopLabelsState:
+    """
+    State for Loop Labels algorithm. Used only with CUDA graphs
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/parts/submodules/rnnt_loop_labels_computer.py]: Added detailed docstrings and commented each variable here and for TDT.

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
jenkins

File: nemo/collections/asr/modules/rnnt.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/collections/asr/modules/rnnt.py:387]: Makes sense

File: nemo/utils/timers.py [hunk 10]
```diff
@@ -158,3 +159,56 @@ def export(self):
         data = {k: fn(v["dt"]) for k, v in self.timers.items() if ("dt" in v)}
 
         return data
+
+
+class SimpleTimer:
```
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/utils/timers.py:164]: Makes sense

**Review by titu1994 — APPROVED** [review]
Looks great !
